### DISCUSSION TITLE
Query builder lacks support of $notExists operator

### DISCUSSION THREAD

**Discussion by thynson** [mikro-orm/mikro-orm#4170] [Q&A]
**Is your feature request related to a problem? Please describe.**
Currently there is no `$notExists` support. And `{$not: { $exists: subQuery }}` does not works, saying that `$exists` is not a column.

**Describe the solution you'd like**
Provide a new operator `$notExists`.

**Describe alternatives you've considered**
Make `{$not: {$exists: subQuery }}` works, through I guess it may be a bit more difficult to support.

**Additional context**

**Comment by B4nan** [comment]
It works fine for me, at least in mongo. Share more details on what does not work for you.

```ts
await em.findOneOrFail(Author, { foo: { $not: { $exists: true } } });
```

Generates following query:

```
[query] db.getCollection('author').find({ foo: { '$not': { '$exists': true } } }, {}).limit(1).toArray(); [took 1 ms]
```

Tried in this test:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/EntityManager.mongo.test.ts#L637-L640

> And {$not: { $exists: subQuery }} does not works

Seeing you mention "subQuery" here, maybe that's the problem - `$exists` expects a boolean value, not a subquery. You can't have anything like subquery in mongo, even if that operator did support it.

**Reply by thynson** [reply]
I'm using it with SQL query builder and expect generated SQL to be something like:

```sql
SELECT * FROM table
WHERE field1=? AND NOT EXISTS ( SELECT * FROM table WHERE field2=?);
```
And I've found a way to work around it by using the method provided by Knex:
```typescript
    const qb1 = this.em.createQueryBuilder(Ticket);
    const qb2 = this.em.createQueryBuilder(Ticket);
    qb1
      .where({ /** some conditions**/})
      .orderBy({serial: 'ASC'})
    qb2.where({ /** some conditions **/}});
    qb1.getKnex().whereNotExists(qb2.getKnex());
```

**Reply by B4nan** [reply]
See, next time you could share more details right ahead, especially when you say something "does not work"...
